File: AddBinary.cpp
#include <iostream>
#include <algorithm>
using namespace std;

class Solution 
{
public:
    string addBinary(string a, string b) 
    {
    	string res = "";
    	int t = 0;
    	int i = a.size() - 1;
    	int j = b.size() - 1;
    	int sum;
    	while (i >= 0 && j >= 0)
    	{
    		sum = (a[i] - '0') + (b[j] - '0') + t;
    		t = sum / 2;
    		sum = sum % 2;
    		res += sum + '0';
    		i--;
    		j--;
    	}
    	while (i >= 0)
    	{
    		sum = (a[i] - '0') + t;
    		t = sum / 2;
    		sum = sum % 2;
    		res += sum + '0';
    		i--;
    	}
    	while (j >= 0)
    	{
    		sum = (b[j] - '0') + t;
    		t = sum / 2;
    		sum = sum % 2;
    		res += sum + '0';
    		j--;
    	}
    	if (t == 1)
    		res += t + '0';
    	reverse(res.begin(), res.end());
    	return res;
    }
};

int main()
{
	Solution s;
	string a = "1";
	string b = "1";
	cout << s.addBinary(a, b) << endl;
	return 0;
}

File: JavaSolution/src/FaceBook/AddBinary.java
package FaceBook;

/**
 * Created by xyunpeng on 1/7/16.
 */
public class AddBinary {
    public String addBinary(String a, String b) {
        if (a == null || a.length() == 0) {
            return b;
        }

        if (b == null || b.length() == 0) {
            return a;
        }

        StringBuilder sa = new StringBuilder(a).reverse();
        StringBuilder sb = new StringBuilder(b).reverse();

        StringBuilder res = new StringBuilder("");
        int next = 0;
        for (int i = 0; i < sa.length() || i < sb.length(); i++) {
            int ad = i >= sa.length()? 0: sa.charAt(i) - '0';
            int bd = i >= sb.length()? 0: sb.charAt(i) - '0';
            res.append(((ad + bd + next) % 2));                         //注意StringBuilder的append是可以自动把int转成char的
            next = (ad + bd + next) / 2;
        }

        if (next != 0) {
            res.append(next);
        }
        return res.reverse().toString();
    }
}


//别人家的代码